format/caff/caff.go

```diff
@@ -0,0 +1,210 @@
+package caff
+
+import (
+	"bytes"
+	"compress/flate"
+	"io"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/bitio"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+var probeGroup decode.Group
+
+func init() {
+	interp.RegisterFormat(
+		format.CAFF,
+		&decode.Format{
+			Description: "Live2D Cubism archive",
+			Groups:      []*decode.Group{format.Probe},
+			DecodeFn:    decodeCAFF,
+			DefaultInArg: format.CAFF_In{
+				Uncompress: true,
+			},
+			Dependencies: []decode.Dependency{
+				{Groups: []*decode.Group{format.Probe}, Out: &probeGroup},
+			},
+		})
+}
+
+const (
+	imageFormatUnknown   = 0x00
+	imageFormatPNG       = 0x01
+	imageFormatNoPreview = 0x7f
+)
+
+const (
+	colorTypeUnknown   = 0x00
+	colorTypeARGB      = 0x01
+	colorTypeRGB       = 0x02
+	colorTypeNoPreview = 0x7f
+)
+
+const (
+	compressOptionRaw   = 0x10
+	compressOptionFast  = 0x21
+	compressOptionSmall = 0x25
+)
+
+var imageFormatNames = scalar.UintMapSymStr{
+	imageFormatUnknown:   "unknown",
+	imageFormatPNG:       "png",
+	imageFormatNoPreview: "no_preview",
+}
+
+var colorTypeNames = scalar.UintMapSymStr{
+	colorTypeUnknown:   "unknown",
+	colorTypeARGB:      "argb",
+	colorTypeRGB:       "rgb",
+	colorTypeNoPreview: "no_preview",
+}
+
+var compressOptionNames = scalar.UintMapSymStr{
+	compressOptionRaw:   "raw",
+	compressOptionFast:  "fast",
+	compressOptionSmall: "small",
+}
+
+type fileInfoListEntry struct {
+	filePath       string
+	startPos       int64
+	fileSize       int
+	isObfuscated   bool
+	compressOption uint8
+}
+
+func decodeVersion(d *decode.D) {
+	d.FieldU8("major")
+	d.FieldU8("minor")
+	d.FieldU8("patch")
+}
+
+func decodeCAFF(d *decode.D) any {
+	var ci format.CAFF_In
+	d.ArgAs(&ci)
+
+	var obfsKey int64
+
+	obfsU8 := func(d *decode.D) uint64 { return d.U8() ^ uint64(obfsKey&0xff) }
+	obfsU32 := func(d *decode.D) uint64 { return d.U32() ^ uint64(obfsKey&0xffff_ffff) }
+	obfsU64 := func(d *decode.D) uint64 { return d.U64() ^ uint64(obfsKey<<32|obfsKey) }
+	obfsBool := func(d *decode.D) bool { return obfsU8(d) != 0 }
+
+	// "Big Endian Base 128" - LEB128's strange sibling
+	obfsBEB128 := func(d *decode.D) (v uint64) {
+		for {
+			x := obfsU8(d)
+			v <<= 7
+			v |= (x & 0x7f)
+			if (x >> 7) == 0 {
+				return
+			}
+		}
+	}
+
+	obfsVarStr := func(d *decode.D) string {
+		length := obfsBEB128(d)
+		if length == 0 {
+			return ""
+		}
+
+		raw := d.BytesLen(int(length))
+		for i := uint64(0); i < length; i++ {
+			raw[i] ^= byte(obfsKey)
+		}
+		return string(raw)
+	}
+
+	d.FieldUTF8("archive_id", 4, d.StrAssert("CAFF"))
+	d.FieldStruct("archive_version", decodeVersion)
+	d.FieldUTF8("format_id", 4)
+	d.FieldStruct("format_version", decodeVersion)
+	obfsKey = d.FieldS32("obfuscate_key")
+	d.SeekRel(8 * 8)
+
+	d.FieldStruct("preview_image", func(d *decode.D) {
+		d.FieldU8("image_format", imageFormatNames)
+		d.FieldU8("color_type", colorTypeNames)
+		d.SeekRel(2 * 8)
```

Comment:
Could this and some of the other `d.SeekRel` be a unused/alignment/padding etc field? i usually try to have no or as few gaps as possibly but i understand it might hard when reversing a format.

BTW when i was talking about "unknown" fields earlier i was confused, they called "gap" fields now, was renamed some time ago so that decoders can use "unknown".

Reply:
I'll change those to `unusedN` fields since I know they're unused here.